### PR TITLE
Complete intro to Post Requests and Database

### PR DESCRIPTION
**Pull request by grenlayk**

### CODE REVIEW

**Comment by grenlayk** [comment]
I tried to fix everything you mentioned.
For now I still have some questions.

**Comment by grenlayk** [comment]
Sometimes (about 1 / 7 times) when I click on "send" button page loads empty, without messages from database, but it supposed to load them 'cause of "onload" in chat.html
Have I wrote something wrong?

**Comment by grenlayk** [comment]
Also, I haven't found the appropriate way to show user ErrorMessage from post request. Can you advise me somehow?

File: portfolio/src/main/java/com/google/sps/data/UserMessage.java [hunk 1]
```diff
@@ -17,32 +17,40 @@
...
     if (userName == null || isEmpty(userName)) {
-      return false;
+      return new UserMessageError("Empty username field");
     }
     if (userMessage == null || isEmpty(userMessage)) {
-      return false;
+      return new UserMessageError("Empty message field");
     }
-    return true;
+    return new UserMessageError("OK");
```
Comment (zajonc-gugle) [inline portfolio/src/main/java/com/google/sps/data/UserMessage.java:53]: I'd suggest simply returning `null` - it's a special case that has to be handled anyway, so let's give it some meaning.

File: portfolio/src/main/java/com/google/sps/data/UserMessage.java [hunk 2]
```diff
@@ -17,32 +17,40 @@
...
 
   public UserMessage(String name, String message) {
-      this.userName = name;
-      this.userMessage = message;
+      this.userName = htmlFilter(name);
+      this.userMessage = htmlFilter(message);
   }
 
-  public boolean isCorrect() {
+  public UserMessage(String name, String message, long id, long timestamp) {
```
Comment (zajonc-gugle) [inline portfolio/src/main/java/com/google/sps/data/UserMessage.java:39]: Since constructors are somewhat magical, I usually pay more attention to them.

In the ideal world, there would exist only one, valid way to initialise an object - a simple, trivial constructor. However, oftentimes some logic has to be performed while building an object (like, for example, calling `htmlFilter` in this case). We have an assumption that `userName` and `userMessage` have been initialized with a value passed through the `htmlFilter` function.

Imagine now, that the signature of `htmlFilter` changes, and another argument is added, say, `bool do_nothing`. We add the new argument at all call sites, passing `do_nothing=false`, except in one place we make a mistake and pass `do_nothing=true`. What happens now, is we notice that the class behaves in a weird way and in some cases doesn't sanitise the input. The problem is, it's not exactly easy to spot the source of the problem - it only surfaces in cases where the constructor is called with 2 arguments instead of 4. It's tricky to find, because reading the code, you don't expect the behaviour to change based on a totally unrelated thing (whether we passed `id` and `timestamp` explicitly or not).

The problem would be avoided if one constructor only added the minimal functionality it must add (the default values) and let the other constructor handle the remaining bits - actual initialisation of the instance.
It's doable by calling `this(args...)` inside the additional constructors: [see stackoverflow](https://stackoverflow.com/questions/285177/how-do-i-call-one-constructor-from-another-in-java)

In short (and this is just an opinion, btw.), I tend to avoid having multiple constructors initialise the class, at all cost. It's a good practice to make similar methods reuse each other, but in the case of constructors it's especially important, because they can do more harm when misused.

### TLDR
- let's initialise the class *only* in the 4-argument constructor
- let's make the 2-argument constructor forward the call to the 4-argument one.

File: portfolio/src/main/java/com/google/sps/servlets/NewMessageServlet.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    String userName = request.getParameter("user_name");
+    String userMessage = request.getParameter("user_message");
+    long timestamp = System.currentTimeMillis();
+
+    UserMessage curMessage = new UserMessage(userName, userMessage);
+
+    if (curMessage.check().getError() != "OK") {
+      System.err.println(curMessage.check().getError());
```
Comment (zajonc-gugle) [inline portfolio/src/main/java/com/google/sps/servlets/NewMessageServlet.java:46]: Regarding your comment about presenting the message to the user:

It would require changing the flow a bit, but I think it's worth it.
In the current implementation, we send a redirect no matter what. In that case, we loose the information about an error - after all, if we return the same result (a redirect response) regardless of what happened in the method, then we cannot communicate what the current state is.

I'd suggest only calling `sendRedirect` inside the `else` clause. In the `if` clause, you can return a different response, e.g. [sendError](https://tomcat.apache.org/tomcat-5.5-doc/servletapi/javax/servlet/http/HttpServletResponse.html#sendError(int,%20java.lang.String)). It will display the message in a text form and not redirect the browser.

If you want to do something more fancy (e.g. display some error in some container, and redirect after some time), that would have to be done on the client side (== in JavaScript). It's totally doable and quite common these days (but I'll spare you the details :) It's probably not too interesting), but it requires some external libraries and a bit of refactoring, so perhaps it's not worth going in that direction (yet).

**Comment by grenlayk** [comment]
According to the problem with /list-messages requests:
Request status in this case is "pending" for it seems like forever. In Timing folder I found "stalled" status. [There](https://developers.google.com/web/tools/chrome-devtools/network/reference?utm_source=devtools#timing-explanation) is an explanation of this status given by Network panel, but I didn't really understood what can it be in may case.

**Comment by grenlayk** [comment]
Also, sometimes I catch same bug (forever pending) with other files (script.js for example).

**Comment by grenlayk** [comment]
Every time I reload page select tag resets to default value. I haven't found simple way to change it, is it ok, or do I need to find way to change it to save user's choice?

**Comment by grenlayk** [comment]
I tested web page once again, just like common user managing through pages and it is also sometimes gets stuck (pending script.js file).

**Comment by zajonc-gugle** [comment]
Sorry for a late response.

Regarding the selected choice - I think it's expected, given the technology we use right now. It's possible to remember user's choices, but that requires the use of cookies and some more fancy frontend logic. I think it's totally fine to leave it as it is.

Regarding stale requests - it seems weird, especially given that it's the `script.js` file that fails to resolve. Since it's effectively a static file, I don't expect our server to fail serving it. I think there's a chance that we're testing Chrome here, rather than the server itself. In general, when debugging such issue, the best first step is to try to find a minimal scenario sufficient to reproduce a bug. In this case, it'd probably require taking Chrome out of the loop and finding a way to fetch all data programatically. Anyway, given that we cannot tell if the problem is caused by our server or by Chrome, I think we can simply live with the problem for now and ignore it, unless it becomes more frustrating.
